### PR TITLE
Update priority admission to improve interoperability

### PR DESCRIPTION
**Pull request by liggitt**
Builds on https://github.com/kubernetes/kubernetes/pull/65722

Makes the following adjustments to the priority admission plugin:
* allows creation of pods to include an explicit priority field if it matches the computed priority (allows export/import cases to continue to work on the same cluster, between clusters that match priorityClass values, and between clusters where priority is unused and all pods get `priority:0`)
* preserves existing priority if a pod update does not include a priority value and the old pod did (allows POST, PUT, PUT, PUT workflows to continue to work, with the admission-set value on create being preserved by the admission plugin on update)

This should avoid the failures revealed by the kubectl tests exercising the pod API without any awareness of the priority feature

/sig scheduling
/cc @bsalamat

```release-note
kube-apiserver: the `Priority` admission plugin is now enabled by default when using `--enable-admission-plugins`. If using `--admission-control` to fully specify the set of admission plugins, the `Priority` admission plugin should be added if using the `PodPriority` feature, which is enabled by default in 1.11.
```

### CODE REVIEW

File: plugin/pkg/admission/priority/admission.go [hunk 1]
```diff
@@ -157,11 +161,21 @@ func (p *priorityPlugin) admitPod(a admission.Attributes) error {
...
+		oldPod, ok := a.GetOldObject().(*api.Pod)
+		if !ok {
+			return errors.NewBadRequest("resource was marked with kind Pod but was unable to be converted")
+		}
+
+		// This admission plugin set pod.Spec.Priority on create.
+		// Ensure the existing priority is preserved on update.
+		if pod.Spec.Priority == nil && oldPod.Spec.Priority != nil {
+			pod.Spec.Priority = oldPod.Spec.Priority
+		}
```
Comment (ravisantoshgudimetla) [inline plugin/pkg/admission/priority/admission.go]: If `pod.Spec.Priority != oldPod.Spec.Priority` shouldn't we throw an error. It means user tried to update priority which is wrong. This field is immutable.

Comment (liggitt) [reply plugin/pkg/admission/priority/admission.go]: API validation already disallows modification. this is preserving the existing value if the incoming pod dropped the field entirely (new pod has it set to nil).

that would occur if a client was using a back-level go struct to get/update, or did `kubectl create -f pod.yaml && kubectl replace -f pod.yaml`

Comment (liggitt) [reply plugin/pkg/admission/priority/admission.go]: the existing test failures on #65722 demonstrate how a client unaware of priority can be bitten by enabling an admission plugin which auto-sets it on create, and doesn't do correspondingly helpful things on update

Comment (ravisantoshgudimetla) [reply plugin/pkg/admission/priority/admission.go]: Thanks @liggitt. One more case - What would happen, if we have 2 priorityClasses with same values. We are setting priority value only here, shouldn't we set priorityClass as well.

Comment (liggitt) [reply plugin/pkg/admission/priority/admission.go]: > What would happen, if we have 2 priorityClasses with same values.

the admission plugin simply ensures the priority value it resolved to is preserved.

if the user attempts to change the priority class name on update, API validation will prevent that as a disallowed change.

**Comment by ravisantoshgudimetla** [comment]
Thanks @liggitt - This LGTM. Let's make sure that @bsalamat sees and approves this.

File: plugin/pkg/admission/priority/admission.go [hunk 2]
```diff
@@ -157,11 +161,21 @@ func (p *priorityPlugin) admitPod(a admission.Attributes) error {
...
-	if operation == admission.Create && pod.Spec.Priority != nil {
-		return admission.NewForbidden(a, fmt.Errorf("the integer value of priority must not be provided in pod spec. Priority admission controller populates the value from the given PriorityClass name"))
+	if operation == admission.Update {
+		oldPod, ok := a.GetOldObject().(*api.Pod)
+		if !ok {
+			return errors.NewBadRequest("resource was marked with kind Pod but was unable to be converted")
+		}
+
+		// This admission plugin set pod.Spec.Priority on create.
+		// Ensure the existing priority is preserved on update.
```
Comment (bsalamat) [inline plugin/pkg/admission/priority/admission.go]: Now that we have this check here, someone who reads the code may not realize that changes to the values are not allowed. I would add to the comment that changes to the priority value or the priority class name are not allowed by the API validation.

**Review by bsalamat — COMMENTED** [review]
Looks good. I just have a minor comment.

Thanks, @liggitt!

File: plugin/pkg/admission/priority/admission.go (hunk 2 above)
Comment (liggitt) [reply plugin/pkg/admission/priority/admission.go]: comment added

**Review by ravisantoshgudimetla — APPROVED** [review]
/lgtm

Thanks @liggitt

**Review by bsalamat — APPROVED** [review]
/lgtm

Thanks, Jordan!

**Comment by liggitt** [comment]
/assign @deads2k 
for admission approval

**Pull request merged**